Doc/regions.md

```diff
@@ -0,0 +1,517 @@
+    frame -->|r| region
+    obj3  -->|f| obj2
+    obj3  -->|g| obj4
+    region-->|f| obj1
+    frame -->|x| obj2
+    frame -->|y| obj3
+    obj1  -->|f| obj2
+    obj2  -->|f| obj1
+    obj2  -->|f| obj4
+    subgraph " "
+        region
+        obj1
+        obj2
+        obj4
+    end
+```
+
+The local reference count for the `region` is modified in the following way.
+* The reference count for `obj2` is increased by 1 to represent the new reference from `obj1`.
+* The objects `obj2` and `obj4` are added to the region, they have a combined reference count of 5.
+* The edge from `obj2` to `obj4` is considered internal, so we reduce the LRC by one.
+* The edge from `obj2` to `obj1` is considered internal, so we reduce the LRC by one.
+* The edge from `obj1` to `obj2` is considered internal, so we reduce the LRC by one.
+
+Thus, the resulting change is to increase the LRC by 5, and then reduce it by 3. 
+
+
+### Immutable objects
+
+To share objects between behaviours, we need to ensure that the object is not mutated.
+This is achieved by making the object "deeply immutable".
+This means that the object and all objects reachable from it are also immutable.
+
+The types in Python are mutable, but are typically only mutated at the start of the program.
+To support this, we need an operation to turn initially mutable objects into immutable objects.
+We extend the programming model with a `freeze` operation that makes an object and its reachable graph of objects immutable.
+
+The fields of an immutable object may not be modified.
+Its reference count can be modified and it can be collected if it is no longer referenced.
+This reference count may be concurrently modified.
+
+To allow globals to be used in a concurrent setting, we add an explicit operation to make the globals immutable.
+```
+make_globals_immutable()
+```
+This must have been run to start the concurrency runtime and enable region creation.
+This cannot be immediately done as we need some time to allow for the set-up of the types and any monkey-patching.
+As types are globals, you would not be able to have any fields on Types (for example default values) if they are immutable from the start.
+
+We do not expect the `make_globals_immutable()` operation to be exposed to the programmer directly, but to be called as needed by the runtime.
+Exposing it in the prototype is useful for testing and debugging.
+
+## Project plan
+
+We need to break the development in to several phases and their completion requirements.
+
+### Phase 1: Deeply immutable objects
+
+*  A flag to represent if an object is deeply immutable.
+*  If the flag is set, then the object cannot be mutated.  (This needs to be enforced in the runtime.)
+*  Any object reachable from a deeply immutable object is also deeply immutable.
+*  A `freeze` operation that makes an object and its reachable graph of objects deeply immutable. 
+*  Reference counts on immutable objects should be atomic instructions.
```

Comment:
I have a feeling I have asked this before but how do we detect that it is time to use atomic instructions? I guess a simple first increment is to make all RC atomic which is not needed in most places, but also not wrong.

Reply:
My thought is, we have a region field in the object.  
* 0 - local object
* 1 - immutable
* ptr - region object/meta-data

We would branch on that in the PyIncRef and PyDecRef code.  If nothing is immutable, then that would be a never taken branch, and no other changes.